File: 0484_find_permutation/find_permutation.go
package main

// Time: O(n)
// Space: O(n)
func findPermutation(s string) []int {
	res := make([]int, len(s)+1)
	for i := 1; i <= len(s)+1; i++ {
		res[i-1] = i
	}
	i := 0
	for i < len(s) {
		j := i + 1
		for j-1 < len(s) && s[j-1] == 'D' {
			j++
		}
		if j-i > 1 {
			swap(res, i, j-1)
		}
		i = j
	}
	return res
}

func swap(arr []int, i int, j int) {
	for i < j {
		arr[i], arr[j] = arr[j], arr[i]
		i++
		j--
	}
}

func main() {

}

File: 0484_find_permutation/find_permutation.py
from typing import List


def find_permutation(s: str) -> List[int]:
    """
    Time: O(n)
    Space: O(1)
    """
    res = [x for x in range(1, len(s) + 2)]
    i = 0
    while i < len(s):
        j = i + 1
        while j - 1 < len(s) and s[j - 1] == "D":
            j += 1
        if j - i > 1:
            swap(res, i, j - 1)
        i = j
    return res


def swap(arr: List[int], i: int, j: int):
    while i < j:
        arr[i], arr[j] = arr[j], arr[i]
        i += 1
        j -= 1
